external-crates/move/crates/move-binary-format/src/compatibility.rs

```diff
@@ -360,69 +362,92 @@ pub enum InclusionCheck {
 }
 
 impl InclusionCheck {
+    pub fn check(&self, old_module: &Module, new_module: &Module) -> PartialVMResult<()> {
+        self.check_with_mode::<InclusionCheckExecutionMode>(old_module, new_module)
+            .map_err(|_| PartialVMError::new(StatusCode::BACKWARD_INCOMPATIBLE_MODULE_UPDATE))
+    }
+
     // Check that all code in `old_module` is included `new_module`. If `Exact` no new code can be
     // in `new_module` (Note: `new_module` may have larger pools, but they are not accessed by the
     // code).
-    pub fn check(&self, old_module: &Module, new_module: &Module) -> PartialVMResult<()> {
-        let err = Err(PartialVMError::new(
-            StatusCode::BACKWARD_INCOMPATIBLE_MODULE_UPDATE,
-        ));
+    pub fn check_with_mode<M: InclusionCheckMode>(
+        &self,
+        old_module: &Module,
+        new_module: &Module,
+    ) -> Result<(), M::Error> {
+        let mut context = M::default();
 
         // Module checks
-        if old_module.address != new_module.address
-            || old_module.name != new_module.name
-            || old_module.file_format_version > new_module.file_format_version
-        {
-            return err;
+        if old_module.name != new_module.name {
+            context.module_name_mismatch(&old_module.name, &new_module.name);
+        }
+
+        if old_module.address != new_module.address {
+            context.module_address_mismatch(&old_module.address, &new_module.address);
         }
 
-        // If we're checking exactness we make sure there's an inclusion, and that the size of all
-        // of the tables are the exact same except for constants.
-        if (self == &Self::Equal)
-            && (old_module.structs.len() != new_module.structs.len()
-                || old_module.enums.len() != new_module.enums.len()
-                || old_module.functions.len() != new_module.functions.len()
-                || old_module.friends.len() != new_module.friends.len())
-        {
-            return err;
+        if old_module.file_format_version > new_module.file_format_version {
+            context.file_format_version_downgrade(
+                old_module.file_format_version,
+                new_module.file_format_version,
+            );
         }
 
+        let mut new_structs_set: HashSet<_> = new_module.structs.keys().collect();
```

Comment:
Generally better to use an ordered set (like `BTreeSet`) for anything that's running in the execution layer. Will help to avoid any unintended forks.

Reply:
Also, is there a way for us to avoid allocating this set and removing elements from it in the case we only care that the sets are different? I.e. can we hand more responsibility to the mode to decide whether it wants to track the actual set of structs/enums/functions that were added, vs just knowing that they are different?

Reply:
since they are already sorted with btreemaps you could do a two pointer where each is simultaneously iterated over to check for the extra/missing, keeps runtime linear, should remove memory overhead.